### PR TITLE
main: slim down projectile pool

### PR DESCRIPTION
**Pull request by chrisforbes**
Signed-off-by: Chris Forbes <chrisf@ijw.co.nz>

### CODE REVIEW

**Comment by mkfifo** [comment]
I like this PR a lot, cannot functionally test until I go home though.

File: main.cc [hunk 1]
```diff
@@ -1065,23 +1054,15 @@ struct play_state : game_state {
...
-                    }
-                }
+            if (num_projectiles < MAX_PROJECTILES) {
+                auto & proj = projectiles[num_projectiles++];
+                proj.pos = pl.eye;
+                proj.dir = pl.dir;
+                proj.velocity = 2.f;
+                proj.lifetime = proj.initial_lifetime;
+
+                pl.fire_projectile = false;
```
Comment (mkfifo) [inline main.cc]: moving this inside the if is an improvement as it means if we could not fire this frame we will fire next frame,
we do have to be careful in our sizing of MAX_PROJECTILES in order to try guarantee that in the unlikely moment we cannot fire this frame we are likely to be able to fire next frame or at the least within a few frames.

File: main.cc [hunk 2]
```diff
@@ -860,25 +849,25 @@ update()
...
-            projectile->draw();
+    /* update the projectiles */
+    for (auto i = 0u; i < num_projectiles;) {
+        if (projectiles[i].update(dt)) {
+            ++i;
         }
         else {
-            available_projectiles.push_back(projectile);
-            current_projectiles[i] = nullptr;
+            projectiles[i] = projectiles[--num_projectiles];
```
Comment (mkfifo) [inline main.cc]: this makes me happy.

File: main.cc [hunk 3]
```diff
@@ -158,11 +152,7 @@ struct projectile
         pos = new_pos;
 
         lifetime -= dt;
-        if (lifetime <= 0.f) {
-            alive = false;
-        } else {
-            alive = true;
-        }
+        return lifetime > 0.f;
```
Comment (mkfifo) [inline main.cc]: +1

**Comment by mkfifo** [comment]
projectile and ray casting functional tests complete, behaves as expected and compares to current master.

**Comment by RobotCaleb** [comment]
what?

**Pull request merged**